modules/nlu/src/backend/application/training-queue.ts

```diff
@@ -0,0 +1,287 @@
+import * as sdk from 'botpress/sdk'
+import * as NLU from 'common/nlu/engine'
+import _ from 'lodash'
+
+import moment from 'moment'
+import ms from 'ms'
+import nanoid from 'nanoid'
+import { ITrainingRepository, ITransactionContext } from './training-repo'
+import { TrainingId, TrainerService, TrainingListener, TrainingState, TrainingSession, I } from './typings'
+
+export interface TrainingQueueOptions {
+  maxTraining: number
+}
+
+const DEFAULT_OPTIONS: TrainingQueueOptions = {
+  maxTraining: 2
+}
+
+const DEFAULT_STATE = { status: <sdk.NLU.TrainingStatus>'idle', progress: 0 }
+
+export type ITrainingQueue = I<TrainingQueue>
+
+const debug = DEBUG('nlu').sub('lifecycle')
+const TRAINING_LIFE_TIME = ms('5m')
+
+const STATES_REQUIRING_NODE_AFFINITY: sdk.NLU.TrainingStatus[] = ['training', 'canceled', 'errored']
+
+export class TrainingQueue {
+  private _options: TrainingQueueOptions
+  private _paused: boolean = true
+
+  private _workerId = nanoid() // TODO: find a way to make this stay between server restart
```

Comment:
use seeded lodash with process.appSecret or nluSeed if it's sticky

Reply:
appSecret is shared between instances, we need the workerID to be unique for each instance of Botpress in the cluster. Using it as a seed to yield the same value for both instances since appSecret is shared across all instances